modules/darwin/general/users.nix

```diff
@@ -0,0 +1,62 @@
+# TODO(low): follow the same structure as NixOS, maybe with a common location?
+
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  sshKeys = [
+    (builtins.readFile (import ../../../external/kalbasit-keys.nix))
+  ];
+
+  makeUser = userName: { uid, isAdmin ? false, home ? "/Users/${userName}" }: nameValuePair
+    (userName)
+    ({
+      inherit home uid;
+
+      gid = 20;
+      isHidden = false;
+      shell = pkgs.zsh;
+    });
+
+  makeHM = userName: { uid, isAdmin, home ? "/Users/${userName}", ... }: nameValuePair
+    (userName)
+    (config.mine.home-manager.config {
+      inherit userName uid isAdmin home;
+      darwinConfig = config;
+    });
+
+  defaultUsers = {
+    yl = { uid = 2000; isAdmin = true; };
+  };
+
+in {
+  options.mine.users = mkOption {
+    type = types.attrs;
+    default = defaultUsers;
+    defaultText = ''
+      The default users are ${builtins.concatStringsSep " " (builtins.attrNames defaultUsers)}
+    '';
+    description = ''
+      The list of users to create.
+    '';
+  };
+
+  config = {
+    users = {
+      knownUsers = builtins.attrNames config.mine.users;
+      knownGroups = [ "mine" ];
+
+      groups = {
+        mine = {
+          gid = 2000;
+          members = ["yl"]; # TODO: This must be computed from the config.mine.users
+        };
+      };
+
+      users = (mapAttrs' makeUser config.mine.users);
+    };
+
+    home-manager.users = mapAttrs' makeHM config.mine.users;
```

Comment:
For some reason, this is not working at all!

# Single user mode:

I'm able to build the configuration with `./scripts/darwin-rebuild build`, but I cannot activate the build!

```console
Activating home-manager configuration for yl
shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
chdir: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
Starting home manager activation
Activating checkFilesChanged
Activating checkLinkTargets
Activating writeBoundary
Activating installPackages
error: opening lock file '/nix/var/nix/db/big-lock': Permission denied
[SHABKA] nix-darwin installer exited with status 1
```

# Multi-user mode:

I reflashed my Mac and started from scratch with a multi-user installation, but I'm getting now an even weirder error:

```console
building '/nix/store/xjl5ynbni3ckns3cmr9pqsmkbjnr2p53-check.drv'...
building '/nix/store/zki2c5vmnf72lin6zh2agqs36mgfzjga-cleanup.drv'...
building '/nix/store/dz13a223lz0l6rqhn5mjkba29m3l56d1-config.nix.drv'...
building '/nix/store/k9nmvhyynbl0wv3p6fjgsx59fh6ppd3s-hm_.gtkrc2.0.drv'...
building '/nix/store/rp6yinjrjlkjw8dh8261r5bz6ch8gsm2-hm_.npmrc.drv'...
building '/nix/store/zjqaghsb4cyjvlg01lf1n0rpab5hm0x9-hm_.taskrc.drv'...
building '/nix/store/n053fknrzbx6599sbzxnrqvmk70jf6b4-hm_.timewarriortimewarrior.cfg.drv'...
building '/nix/store/rr1fikhm1mc9rz670iasywn69vw1hnwq-hm_.zshenv.drv'...
building '/nix/store/dghcziw4g4hhvylxaryc6s5ddimkrqcy-hm_Usersyl.localsharetask.keep.drv'...
building '/nix/store/zmil7zp0gwwypivh0fhlip7fl5kljw97-hm_fontconfigconf.d10nixprofilefonts.conf.drv'...
error: all build users are currently in use; consider creating additional users and adding them to the 'nixbld' group
```

Removing this line allows me to build and switch!

@LnL7 @rycee I'm not sure why this is happening, thoughts?

Reply:
I have a very vague understanding of how Home Manager and nix-darwin interacts. Perhaps @ElvishJerricco have some idea?

Regarding the single-user case, one of the first things happening in the HM activation script is `cd $HOME` so if `$HOME` is set incorrectly or the user home directory doesn't exist then I can imagine you getting the chdir error. The HM nix-darwin module runs the HM activation script in `system.activationScripts.extraActivation.text` and looking at

https://github.com/LnL7/nix-darwin/blob/166560ca767e98b6f2f3b23a2f4dadb9f849a532/modules/system/activation-scripts.nix#L55-L57

I note that `system.activationScripts.users.text` seems to run _after_ `extraActivation`, which I guess mean that the user would be created after HM is run. So perhaps the HM activation script should run in `system.activationScripts.postActivation.text` or something instead?

Reply:
The nix-darwin interaction [is extremely basic](https://github.com/rycee/home-manager/blob/master/nix-darwin/default.nix). All it does is switch to each user and run the home-manager activation script for that user, during nix-darwin activation. Not really sure how that could be causing problems.

Reply:
1. looks like a user other than the store owner is running nix commands.
2. Either it's not a proper multi-user install (install --daemon) or nix.maxJobs / nix-build -j is higher than the number of nixbld users available.

Reply:
@LnL7 I have 32 nixbld users (I checked with `id nixbldN`). Here's my nix.conf:

```
build-users-group = nixbld

max-jobs = 32
cores = 1
sandbox = false
```

I think it's all happening because of the way nix-darwin is invoking home-manager by attempting to run as the target user. Is there any other way to run the home-manager activation script @ElvishJerricco?

Reply:
> I note that `system.activationScripts.users.text` seems to run _after_ `extraActivation`, which I guess mean that the user would be created after HM is run. So perhaps the HM activation script should run in `system.activationScripts.postActivation.text` or something instead?

This explains the issue in my TODO. I was running into this issue in single user mode as well. I'll submit a PR for it.

Reply:
I re-installed again. This time in single user mode and home-manager is patched with https://github.com/rycee/home-manager/pull/583 and now I'm getting a similar but slightly different error:

```console
athena:shabka admin$ ./scripts/darwin-rebuild.sh switch
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
+ darwin-rebuild -I nixpkgs=/nix/store/pfrdp2fx1f3vw9v1jfznxnag9v5h901l-nixpkgs-unstable-36f316007494c388df1fec434c1e658542e3c3cc -I darwin=/nix/store/i71wsp1zr1i60pqwqwg5hppizapc1hx5-nix-darwin-10c34f12775dd941ecb796f426272f891bde5ec7 -I darwin-config=hosts/athena/configuration.nix switch
building the system configuration...
user defaults...
setting up user launchd services...
creating user service org.nixos.gnupg-agent
setting up groups...
creating group mine...
updating group members mine...
setting up users...
creating user yl...
setting up ~/Applications...
setting up /etc...
warning: not linking environment.etc."zshrc" because /etc/zshrc exists, skipping...
system defaults...
setting up launchd services...
reloading service org.nixos.activate-system
configuring time...
configuring networking...
configuring keyboard...
Activating home-manager configuration for yl
Starting home manager activation
Activating checkFilesChanged
Activating checkLinkTargets
Activating writeBoundary
Activating installPackages
error: opening lock file '/nix/var/nix/db/big-lock': Permission denied
```

As you can see @rycee, there was no chdir error, but it's definitely an error grabbing the lock. Is it failing at `nix-env -i` stage?

Reply:
It seems to me that this is the same problem as with running [home-manager through a VM](https://github.com/rycee/home-manager/blob/81dae2f88eaff306f36b9072d2cb752f82dbc60a/modules/home-environment.nix#L313-L340). @LnL7 does nix-darwin support the concept of `users.users.<name?>.packages`?